Quantized-mesh output support

ahuarte47 opened the pull request:
Provides some new features:
- New quantized-mesh format ouput (-f Mesh), optionally you could add VertexNormals with -N option.
- New metadata file output.
- New cesium-friendly option to automatically create missing root tiles (level 0).
- Consider nodata values when creating tiles.
- Catch GDAL 'Integer overflow' error when creating tiles of low levels.
- Refactoring of code to easily write new output formats.
  
Details:
https://www.linkedin.com/pulse/fast-cesium-terrain-rendering-new-quantized-mesh-output-alvaro-huarte/

kvuorine commented:
Seems very good so far during my tests, well done.

Performance of Cesium took a leap forward compared to heightmap terrain tiles.

There is one problem, when I tried to play with --mesh-qfactor option, ctb-tile segfaulted.

ahuarte47 commented:
Thanks @kvuorine. About --mesh-qfactor error, may I have your settings or data to try fix it?

kvuorine commented:
With commands below you can get some of the data I am working with, I tested that this dataset produces 
--mesh-qfactor error too
```Shell
wget http://kartat.kapsi.fi/files/korkeusmalli/hila_2m/etrs-tm35fin-n2000/L4/L41/L4141E.tif http://kartat.kapsi.fi/files/korkeusmalli/hila_2m/etrs-tm35fin-n2000/L4/L41/L4141G.tif http://kartat.kapsi.fi/files/korkeusmalli/hila_2m/etrs-tm35fin-n2000/L4/L41/L4141F.tif http://kartat.kapsi.fi/files/korkeusmalli/hila_2m/etrs-tm35fin-n2000/L4/L41/L4141H.tif
gdalbuildvrt merged.vrt L4141?.tif
mkdir terrain
ctb-tile -o terrain -f Mesh -g 1.1 merged.vrt
```

ahuarte47 commented:
ok, I'll review this as soon as possible

ahuarte47 commented:
Hi @kvuorine, I have fixed this option. Thanks for your notice

Rik79 commented:
Hello!  I tried to use this code to generate quantized mesh for Cesium and it works very well.
Thank you! :+1:  :1st_place_medal: 

I had only an issue when i tried to generate tiles from two different adjacent tiff images (from SRTM dataset) : on the seam between these tiles there is a "canyon", a crack. I am sure that the tiffs data is correct, they are attached.

I launched "ctb-tile" 2 times, each with a tile and on the seam this is the result...

do you have some advice? the two tiles should be connected before launch ctb-tile?

The result...

<img src='http://paraglidinglogbook.com/share/crack.jpg'>

ahuarte47 commented:
Hi, I think you need an unique MDT that covers all area in order to execute the ctb-tile only once. You could create a virtual raster with "gdalbuildvrt" to virtually create an unique source from many individual files.

You can see a previous comment to how to use or read about "gdalbuildvrt" tool:

```
gdalbuildvrt merged.vrt L4141?.tif
mkdir terrain
ctb-tile -o terrain -f Mesh -g 1.1 merged.vrt

```

Rik79 commented:
Ok thank you very much, i will try this command! (Probably this is the solution, but i am asking the need of memory for the entire world tileset!)

Looking at the image is interesting to note that the first generated tileset on the right is "cutted", the second on the left is "seamed" to globe. Futhermore looking far where the tiles are loaded with minor zoom level the canyon appers to be larger.  These can be clues to uderstand.
I am wondering if "ctb-tile" has an option not to seam to the ground level the tileset.

About the generation of the "layout.json" file, also this  works very well, but again in my case with two tiles generated it is not very usefull. Could be interesting a tool that analize the directory tree and from this generate the layout.json.

All this things let me understand that my approach is not correct. :)

Rik79 commented:
Ok, as not said. :)

"gdalbuildvrt" solves perfectly the problems, also the "layer.json" one. Thank you a lot! :)

ediebold commented:
Is there a technical reason as to why the new "-l" option is separate from the actual terrain generation? For large datasets, generating the layer.json file still takes quite a while.

ahuarte47 commented:
Hi @ediebold, there is none technical reason. But I thought it seems better two separate tasks, I can wrong but it looks more versatile. You could create the layer.json of old terrain sets, or update only a specific level without overwrite an existing json file (You could have edited it with your own settings). If you update a level, only the limits of it are calculated.

kvuorine commented:
How about creating layer.json if there is none when generating tiles, but update layer.json only with command line option selected?

I have been testing --mesh-qfactor option which now doesn't crash the program, but I don't quite get the purpose of this option because tileset size on disk doesn't seem to vary at all when testing different values. @ahuarte47 could you explain it a bit?

ahuarte47 commented:
Thanks everybody for your advices, I will try to integrate the -l option as complementary to -f option, but it will require me to do many changes in the original base code, but indeed it seems more logic.

About _--mesh-qfactor_ option, it is supposed larger values should mean minor amount of triangles with minor quality. As my little [tutorial ](https://www.linkedin.com/pulse/fast-cesium-terrain-rendering-new-quantized-mesh-output-alvaro-huarte/)said, the process takes care of an input geometric error to safely merge each pair of triangles with its parent one if there is not a visible loss of quality. The factor multiplies the default error value, larger values should mean minor quality. Did not you notice any change?

ahuarte47 commented:
Hi @kvuorine, I changed the code, now CTB creates layer.json if there is none when generating tiles. Thanks everybody!

kvuorine commented:
@ahuarte47 Thanks for your support.

I did notice difference in terrain. I just thought simplified mesh would take less space on disk and be quicker to serve over net. I guess simple mesh still reduces rendering workload.

I tried to read your docs/tutorials earlier but that registration barrier was just too much for me.

epifanio commented:
How to serve the generated tiles? @ahuarte47 I guess storing them on a simple apache directory is not an option, right? I tried to copy some tiles here [temp-tiles](http://epinux.com/epinux_data/shared/data/terrain2/) (they are using the same files used by @kvuorine )
[sand-castle](https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/index.html#c=dVLbbtswDP0VIi9xAU9K0rVoMycY0D1mW7EFfTIwMDYdC5WlTJLdZkP/ffQlaZt29IMliucc3hp0EMg5VObW2Ubl5GABhh7ghryqK9H/1q9Dor+pAbbaaZjDuAxh5+dS0k6Z+lFkthqOv3IMKH2JjnLZnQetmRzHPYWj3zX5cEcu0OM36yrUnjmDqyk1T2efUtN+DafZKHo4ze6u80XjrLveWBOYnNw4hiHDDXpa4Z7crcruGT2HggVoED+tfH7qOaYgJaxLgq22G4Kq9gHI4EYTaLUtgzJbCBYqvCeoPYEtIHD0QDb20HTlgenr43q6vIXPyJDoSEXPtzrQLboWDOUfARlW5FAUer+2hyG0lnMLlcGgrOEiDqNDFvUKzbkonK2+0NYR+ejD5ZWYTKez6flVb9cxfJyy63J2cbQYLiYTMTmLnzWsU8T9HTReiLdWEuZt2kfxrxhKEewPdqPx0QlXazsVsvK/gOt3EM7qdt/45dn/1B+HMY3iUeLDXtOyd39W1c660C5qJARvX7XTyE2Rm5rXIYjM+xaYyAMoyVUDKl+ko5OVSkeQafSeX4pa65/qD6WjZSI5/hVM264R33niGvdtSDldrnqnECKRfH2LCtbqDboXjP8A) example returns: 

```
An error occurred in "CesiumTerrainProvider": An error occurred while accessing https://epinux.com/epinux_data/shared/data/terrain2/layer.json.
```

ahuarte47 commented:
Set as URL the root directory: 
https://epinux.com/epinux_data/shared/data/terrain2/ 

without the json file.

ahuarte47 commented:
You have clone the missing level-0 folder from the other existing one (You can google about it, you will find this trick). 

https://github.com/geo-data/cesium-terrain-builder/issues/2

kvuorine commented:
@epifanio , I also have following apache directives for the directory
```
AddType application/octet-stream terrain
AddEncoding gzip terrain
```
Not sure if quantized-mesh requires them, but I had problems serving heightmap tiles without those.

muratkendir commented:
Hi,
I finally rendered our first city-level quantized-mesh terrain with help of Alvaro's CTB. Thanks for all kind of useful information.

![2018-03-20 10_55_52-cesium demo_sm](https://user-images.githubusercontent.com/3502531/37821056-9a5e4160-2e93-11e8-9fb9-09fa3f100bb4.jpg)

ahuarte47 commented:
I think you need the other root terrain tiles.
https://epinux.com/epinux_data/shared/data/terrain/0/1/0.terrain

You should have:
https://epinux.com/epinux_data/shared/data/terrain/0/0/0.terrain
https://epinux.com/epinux_data/shared/data/terrain/0/1/0.terrain

epifanio commented:
@ahuarte47 I moved the tiles inside the cesium app so to make them "available" to cesium without using apache (I thought it was a correct way to avoid apache .htaccess (which I also tried)). I do have my app running on 8000 same host at the moment.  I'll try again with apache as well.

ahuarte47 commented:
I don't have this "...Apps\tiles\0\0\1.terrain"

ahuarte47 commented:
I load your terrain with this root folder:
[0.zip](https://github.com/geo-data/cesium-terrain-builder/files/1841523/0.zip)

epifanio commented:
I see, I manually added it thinking that's was a missing tiles , my several attempts to debug.. 

removing it this is the log I have:

in the html I point to the tiles directory (no apache):

```
    var terrainProvider = new Cesium.CesiumTerrainProvider({
      url: 'tiles/'
    });
```

Log:

```
[Log] An error occurred in "w": Failed to obtain terrain tile X: 0 Y: 0 Level: 0. (Cesium.js, line 477)
[Log] An error occurred in "w": Failed to obtain terrain tile X: 1 Y: 0 Level: 0. (Cesium.js, line 477)
```

Using apache dir I have a cross origin problem (I'm googling about it):

html code:
```
    var terrainProvider = new Cesium.CesiumTerrainProvider({
      url: 'https://epinux.com/epinux_data/shared/data/tiles/'
    });
    viewer.terrai
```
Log:
```
[Error] Origin http://epinux.com:8000 is not allowed by Access-Control-Allow-Origin.
[Error] Failed to load resource: Origin http://epinux.com:8000 is not allowed by Access-Control-Allow-Origin. (layer.json, line 0)
[Error] XMLHttpRequest cannot load https://epinux.com/epinux_data/shared/data/tiles/layer.json due to access control checks.
[Log] An error occurred in "w": An error occurred while accessing https://epinux.com/epinux_data/shared/data/tiles/layer.json. (Cesium.js, line 477)
```

I tried to change ` https://epinux.com:8000` with ` http://localhost:8000` but then the log says:

```
[Log] An error occurred in "w": An error occurred while accessing https://localhost/epinux_data/shared/data/tiles/layer.json. (Cesium.js, line 477)
[Error] Failed to load resource: An SSL error has occurred and a secure connection to the server cannot be made. (layer.json, line 0)
```

finally removing `https` and using `http`, the log is:

```
[Error] Failed to load resource: Could not connect to the server. (layer.json, line 0)
```

I can leave the process running if usual to debug

ahuarte47 commented:
Sorry, these errors are related with Cesium deployment. I only can say a last try, edit your "web.config" in Cesium root folder adding this:

```
<httpProtocol>
  <customHeaders>
      <add name="Access-Control-Allow-Origin" value="*" />
      <add name="Access-Control-Allow-Headers" value="origin, x-requested-with, content-type" />
      <add name="Content-Encoding" value="gzip" />
  </customHeaders>
</httpProtocol>
```

epifanio commented:
That's a bit frustrating, I started from scratch on my laptop (before was on a remote server) 
I can run Cesium with its own server.js or just storing the Cesium directory under the apache root dir.
I can run all the examples. In having Cesium served via apache I do not have the control origin error (both "tiles" and "cesium app" are served by apache on localhost on my laptop) but the tiles are giving me the same error:

```
[Log] An error occurred in "w": Failed to obtain terrain tile X: 0 Y: 0 Level: 0. (Cesium.js, line 477)
[Log] An error occurred in "w": Failed to obtain terrain tile X: 1 Y: 0 Level: 0. (Cesium.js, line 477)
```

btw ... this log is not very useful.

the only thing left to try is to add the `.htaccess` inside the root `/tiles/` directory.
I'm gonna try with this `.htaccess`:

```
Header add Access-Control-Allow-Origin "*"
Header add Access-Control-Allow-Headers "origin, x-requested-with,
content-type"

<IfModule mod_headers.c>
  <FilesMatch "\.(png|jpeg)$">
    SetEnvIf Origin ":" IS_CORS
    Header set Access-Control-Allow-Origin "*" env= IS_CORS
  </FilesMatch>
</IfModule>

AddType application/octet-stream terrain
AddEncoding gzip terrain
```

epifanio commented:
no luck, actually trying to load the page with chrome instead of safari, the cesium widget doesn't load at all with a nice message:

![screen shot 2018-03-23 at 3 46 27 pm](https://user-images.githubusercontent.com/18949/37835834-602d59c2-2eb1-11e8-81ac-3bfcf04a7702.png)

webgl works because the src examples (e.g.: HelloWorld) works fine.

I'm gonna write on the mailing list to seek help, I already bloated the whole thread here :) 
Huge thanks for the help!

markerikson commented on tools/ctb-tile.cpp:
```diff
@@ -451,6 +690,14 @@ runTiler(TerrainBuild *command, Grid *grid) {
 
+  // Pass metadata to global instance.
+  if (threadMetadata) {
+    static std::mutex mutex;
```
I could be wrong, but this seems broken. We're creating a new mutex instance in every thread, rather than having all threads share the same global mutex.

This example looks relevant: http://www.bogotobogo.com/cplusplus/C11/7_C11_Thread_Sharing_Memory.php

ahuarte47 replied:
But, it is declared as static, I think it is shared by all threads

ahuarte47 replied:
Am I wrong?

markerikson replied:
Ah, that might work.  Sorry, my C++ is pretty shaky :)

I'm actually trying to use your branch as a starting point for implementing MBTiles support in CTB, so I'm having to dust off my C++ skills for the first time in years.

ahuarte47 replied:
Don't worry. You are welcome! umm MBTiles support! it looks great @markerikson

ediebold commented:
did something change here recently? I was able to build fine a week ago, but now I am encountering issues. I'm using GDAL 2.1.3

ahuarte47 commented:
This [change](https://github.com/geo-data/cesium-terrain-builder/commit/3b9394fa0f82980ed62953ae9c0666231a24fe00) was done in parent branch.

ediebold commented:
Huh, didn't realise the original repo was even still active. Okay, I'll go upgrade my GDAL

ahuarte47 commented:
Hi @ediebold, I did a minor change, do you can build now?

ediebold commented:
I upgraded my GDAL to 2.2.2 - but have hit a new issue (with and without that minor change, so definitely a 2.2 issue).

Compiler message is:

../src/libctb.so: undefined reference to `GDALCreateOverviewDataset(GDALDataset*, int, int)'
collect2: error: ld returned 1 exit status
tools/CMakeFiles/ctb-tile.dir/build.make:98: recipe for target 'tools/ctb-tile' failed
make[2]: *** [tools/ctb-tile] Error 1
CMakeFiles/Makefile2:202: recipe for target 'tools/CMakeFiles/ctb-tile.dir/all' failed
make[1]: *** [tools/CMakeFiles/ctb-tile.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2

markerikson commented:
Yeah, my understanding is that that method is no longer even being exported in GDAL 2.2.  The signature was also changed, but it's not being exported, so you can't link to it.  (This discussion should probably be moved over to #15 .)

ediebold commented:
I am testing the quantized mesh output now- but just to confirm, do we still need to add any missing base tiles ourselves? Is this something we could make ctb-tile do automatically? This is really the only thing preventing the automatic hosting of ctb-tile output.

ediebold commented:
Also, I've done some tests, and this branch works fantastic for me.

markerikson commented:
Yeah, my initial tests with a couple different source terrain datasets show it seems to be working well.

I'm actually hoping to write up a blog post after I've had a chance to generate a complete dataset.

ediebold commented:
Okay, I've found one problem.

It seems that the layer.json's available array isn't being filled out for all the layers. I am processing a dataset that goes down to zoom level 16. However, when viewed with the generated layer.json file, I can't get past zoom level 12. This is true for both quantized mesh and heightmap. When I build a heightmap tileset and delete the entire available entry from the json, I can see all the levels of detail.

Edit: I added the missing layer's availabiltiy to the json file manually, and it fixed the problem. Of my 17 zoom levels (0 included), only 13 had entries in the available list initially.

Edit 2: Just to confirm, running with the -l flag doesn't help.

ediebold commented:
Okay, I've hit another issue. Not 100% sure this belongs here, but this place seems to be where problems get fixed.

I've made a VRT which is a combination of two SRTM 30m files, and a DEM we collected ourselves. I used gdalwarp and gdal_translate to make sure all three files had the same projection and band type.

Everything goes fine for the first 10 levels of depth (level 16 till 7), but then I get this error message:

```
ERROR 1: Integer overflow : nSrcXSize=22062, nSrcYSize=30960
ERROR 1: IReadBlock failed at X offset 0, Y offset 0
ERROR 1: GetBlockRef failed at X block offset 0, Y block offset 0
Error: Could not read heights from raster
ERROR 1: Integer overflow : nSrcXSize=22268, nSrcYSize=31465
ERROR 1: IReadBlock failed at X offset 0, Y offset 0
ERROR 1: GetBlockRef failed at X block offset 0, Y block offset 0
Error: Could not read heights from raster

```
This leaves me with no tiles at a zoom level lower than 7 (i.e. I have tiles from 16 to 7). The folder structure for zoom level 6 exists, but is empty. I know this is a VRT issue since the individual files all work fine

Edit: Interestingly, I made another VRT out of the files, but set the resolution to "lowest", rather than "highest", as it was before. This seems to prevent the issue, but obviously isn't a solution since I'm after the higher resolution of the other DEM. Setting the resolution to "average" also fixes the problem, but also isn't what I'm after.

Edit 2: Even though the process fails out early, it still produces a layer.json file. Everything seems fine, except for the available list, which is obviously missing the early zoom levels. Interestingly, it has the details for level 6, the level it is failing out on:

```
"available": [
    [  ]
   ,[  ]
   ,[  ]
   ,[  ]
   ,[  ]
   ,[  ]
   ,[ { "startX": 114, "startY": 29, "endX": 114, "endY": 29 } ]
   ,[ { "startX": 228, "startY": 59, "endX": 229, "endY": 60 } ]
   ,[ { "startX": 457, "startY": 118, "endX": 459, "endY": 120 } ]
   ,[ { "startX": 915, "startY": 236, "endX": 918, "endY": 241 } ]
   ,[ { "startX": 1831, "startY": 472, "endX": 1837, "endY": 483 } ]
   ,[ { "startX": 3663, "startY": 944, "endX": 3675, "endY": 967 } ]
   ,[ { "startX": 7327, "startY": 1888, "endX": 7350, "endY": 1934 } ]
  ]
```
Without being familiar with the code, this suggests to me that the issue is separate from the layers.json construction at least.

markerikson commented:
@ediebold : the "Integer Overflow" issue appears to be the same thing as #28 and https://github.com/geo-data/cesium-terrain-builder/issues/37#issuecomment-252857173 .

Another option besides building multiple VRT levels may be using `gdaladdo` to add overviews for your dataset.

ediebold commented:
I just came here to say I'd found the answer, but you beat me to it.

I've done the fix from #37 , but it's worth noting that doing so breaks the layer.json creation- since layer.json is only created if one doesn't already exist. To get around this, you need to re-call ctb-tile with the "-l" flag. Of course, at the moment, this still runs into my first issue - I don't get all the zoom levels in the layers.json file.

ahuarte47 commented:
Both issues (Missing root tile for CesiumJS, and "Integer Overflow" error) are old known issues but not related with Quantized-mesh output.

I could fix some of them (first one looks easy to develop), but I have the doubt if it is better to create specific pull requests in order to do not mix features. Any thoughts?

kvuorine commented:
@ahuarte47 I think that Missing root tile issue is not a bug in ctb-tile but Cesium should be fixed to handle this.

I see no sense in generating empty tiles for terrain or imagery.

I opened an issue https://github.com/AnalyticalGraphicsInc/cesium/issues/6301 for this while ago but closed it when I found that "missing" root tile was causing my problem with only half of the globe rendered when no terrain tile for the other half. I'm thinking of reopening it and explaining myself better.

Edit: Cesium already handles this for imagery tiles.

ediebold commented:
@ahuarte47 - I agree about the scope of this pull request. However, this main repo has been effectively dead for years. Would it be better if we just began raising issues in your fork?

Also, what about the issue with layers.json? That would almost certainly be an issue for this pull request?

ahuarte47 commented:
Hi @kvuorine 

> I think that Missing root tile issue is not a bug in ctb-tile but Cesium should be fixed to handle this.

I agree, but does it have sense add a new option to workaround this issue now?

> I opened an issue AnalyticalGraphicsInc/cesium#6301 for this while ago but closed it when I found that "missing" root tile was causing my problem with only half of the globe rendered when no terrain tile for the other half. I'm thinking of reopening it and explaining myself better.

+1

ahuarte47 commented:
Hi @ediebold 

> I agree about the scope of this pull request. However, this main repo has been effectively dead for years. Would it be better if we just began raising issues in your fork?

I am going to do my best, but I can not offer official support for this.

> Also, what about the issue with layers.json? That would almost certainly be an issue for this pull request?

I think the best option is to do not create the layer.json if any error occurs and no all scales are specified when executing the tool. it seems congruent me.

ediebold commented:
@ahuarte47

I think you misunderstood which issue I’m talking about. I meant this comment:

https://github.com/geo-data/cesium-terrain-builder/pull/55#comment-376384683

It seems that the layer.json’s availability list is required for quantized mesh terrains, so I can’t not include it. Based on the fact that only a few zoom levels are missing, I’m hoping it’s just a typo somewhere.

ahuarte47 commented:
Sorry, I am going to try explain my point, only the json file will be created if -l option or all levels are specified (from 0 to....).

`ctb-tile    -s 16 -e 5 -p geodetic -f Terrain -o C:\folder raster.tif ` 
No create the json file although it does not exist.

`ctb-tile    -s 16 -e 0 -p geodetic -f Terrain -o C:\folder raster.tif ` 
Create the json file if it does not exist.

`ctb-tile -l -s 16 -e 5 -p geodetic -f Terrain -o C:\folder raster.tif` 
`ctb-tile -l -s 16 -e 0 -p geodetic -f Terrain -o C:\folder raster.tif` 
Always trying to create the json file.

-l option does not open any input raster file, only travels by all tiles to count them. You can always execute it with all levels.

If you get some "Integer Overflow" error, you would have to execute, by example:

`ctb-tile -s 16 -e 7 -p geodetic -f Terrain -o C:\folder raster_05m_resolution.tif` 
`ctb-tile -s   6 -e 0 -p geodetic -f Terrain -o C:\folder raster_10m_resolution.tif` 
`ctb-tile -s 16 -e 0 -p geodetic -f Terrain -o C:\folder raster_5m_resolution.tif -l` 

Edit:
In order to avoid confusion, I think it is better to create the json file only when -l option is specified. "Integer overflow" error will not create more corrupt json files. And also, this step will be more fast than previous versions because of it take advantage of available CPU/Threads. ok?

ahuarte47 commented:
I added a new "cesium-friendly" option to force the creation of the missing root tile when the 0-level is included in the command settings. It copies the existing file to the missing location. If everybody think that is bad option, I rollback this commit.

ediebold commented:
@ahuarte47 

My problem is that running

`ctb-tile -s 16 -e 0 -p geodetic -f Terrain -o C:\folder raster.tif`

Still only creates a layers.json file that has layers 0-12 in the availability. There is no entry for layers 13-16, even though the tiles exist. There's not even empty brackets. This is separate from my other problems, and exists on the tif file alone, no VRT.

Edit: And running -l afterwards has the same problem. I never get a layers.json file that covers those layers. Seems to be a bug in how the layers.json's availability is filled in.

ahuarte47 commented:
OK, I can not reproduce it with my data, could you test adding "--thread-count 1", if it runs right, it gives me a clue. Thanks!

ediebold commented:
Huh,

Setting the thread count to one actually made it worse- now I only get layers 0 to 8.

ahuarte47 commented:
Could I have your raster to test?

ediebold commented:
I'll double check I'm okay to share it.

In the meantime, this is what I got when I set the thread count to 4:

```
...
  "available": [
    [ { "startX": 1, "startY": 0, "endX": 1, "endY": 0 } ]
   ,[ { "startX": 3, "startY": 0, "endX": 3, "endY": 0 } ]
   ,[ { "startX": 7, "startY": 1, "endX": 7, "endY": 1 } ]
   ,[ { "startX": 14, "startY": 3, "endX": 14, "endY": 3 } ]
   ,[ { "startX": 28, "startY": 7, "endX": 28, "endY": 7 } ]
   ,[ { "startX": 57, "startY": 14, "endX": 57, "endY": 14 } ]
   ,[ { "startX": 114, "startY": 29, "endX": 114, "endY": 29 } ]
   ,[ { "startX": 229, "startY": 59, "endX": 229, "endY": 59 } ]
   ,[ { "startX": 458, "startY": 119, "endX": 458, "endY": 119 } ]
   ,[ { "startX": 917, "startY": 238, "endX": 917, "endY": 238 } ]
   ,[ { "startX": 1834, "startY": 477, "endX": 1835, "endY": 477 } ]
   ,[ { "startX": 3668, "startY": 954, "endX": 3670, "endY": 956 } ]
   ,[  ]
   ,[  ]
   ,[  ]
   ,[  ]
  ]
}
```

Seems like you might be right about it being a threading issue.

ahuarte47 commented:
I tried a change, could you test it? I would greatly appreciate it

ediebold commented:
Edit: Just saw your change, will try it now.

Okay, I can't share the one I've been using, but the good news is that I've got it to fail on standard SRTM data.

Using [this site](http://dwtkns.com/srtm30m/), I managed to download the [SRTM tile S07E142](https://e4ftl01.cr.usgs.gov//MODV6_Dal_D/SRTM/SRTMGL1.003/2000.02.11/S07E142.SRTMGL1.hgt.zip)

Then I ran:

```
ctb-tile -o tiles -f Mesh -C S07E142.hgt
ctb-tile -o tiles -f Mesh -C -l S07E142.hgt
```

The output is tiles for zoom levels 0-14. However, my layers.json file looks like this:

```
{
  "tilejson": "2.1.0",
  "name": "S07E142",
  "description": "",
  "version": "1.1.0",
  "format": "quantized-mesh-1.0",
  "attribution": "",
  "schema": "tms",
  "tiles": [ "{z}/{x}/{y}.terrain?v={version}" ],
  "projection": "EPSG:4326",
  "bounds": [ 0.00, -90.00, 180.00, 90.00 ],
  "available": [
    [ { "startX": 1, "startY": 0, "endX": 1, "endY": 0 } ]
   ,[ { "startX": 3, "startY": 0, "endX": 3, "endY": 0 } ]
   ,[ { "startX": 7, "startY": 1, "endX": 7, "endY": 1 } ]
   ,[ { "startX": 14, "startY": 3, "endX": 14, "endY": 3 } ]
   ,[ { "startX": 28, "startY": 7, "endX": 28, "endY": 7 } ]
   ,[ { "startX": 57, "startY": 14, "endX": 57, "endY": 14 } ]
   ,[ { "startX": 114, "startY": 29, "endX": 114, "endY": 29 } ]
   ,[ { "startX": 228, "startY": 59, "endX": 229, "endY": 59 } ]
   ,[ { "startX": 457, "startY": 118, "endX": 459, "endY": 119 } ]
   ,[ { "startX": 915, "startY": 236, "endX": 918, "endY": 238 } ]
   ,[ { "startX": 1831, "startY": 472, "endX": 1837, "endY": 477 } ]
   ,[ { "startX": 3663, "startY": 944, "endX": 3675, "endY": 955 } ]
  ]
}
```

Also, when I ran the "-l" version of the command, this is what was printed on stdout:

`0...10...20...30...40...50...60...70...80...90...0...10...20...30...40...50...60...70...80...90...100 - done.`

Also, when I run with "--thread-count 4", I get the following in stdout:


```
0...10...20...30...40...50...60...70...80...90...100 - done.
0...10...20...30...40...50...60...70...80...90...100 - done.
```

And my layers.json looks like this:

```
{
  "tilejson": "2.1.0",
  "name": "S07E142",
  "description": "",
  "version": "1.1.0",
  "format": "quantized-mesh-1.0",
  "attribution": "",
  "schema": "tms",
  "tiles": [ "{z}/{x}/{y}.terrain?v={version}" ],
  "projection": "EPSG:4326",
  "bounds": [ 0.00, -90.00, 180.00, 90.00 ],
  "available": [
    [ { "startX": 1, "startY": 0, "endX": 1, "endY": 0 } ]
   ,[ { "startX": 3, "startY": 0, "endX": 3, "endY": 0 } ]
   ,[ { "startX": 7, "startY": 1, "endX": 7, "endY": 1 } ]
   ,[ { "startX": 14, "startY": 3, "endX": 14, "endY": 3 } ]
   ,[ { "startX": 28, "startY": 7, "endX": 28, "endY": 7 } ]
   ,[ { "startX": 57, "startY": 14, "endX": 57, "endY": 14 } ]
   ,[ { "startX": 114, "startY": 29, "endX": 114, "endY": 29 } ]
   ,[ { "startX": 228, "startY": 59, "endX": 229, "endY": 59 } ]
   ,[  ]
   ,[  ]
   ,[  ]
   ,[  ]
   ,[  ]
   ,[ { "startX": 14654, "startY": 3777, "endX": 14693, "endY": 3822 } ]
  ]
}
```

ahuarte47 commented:
Thanks, I am going to test these data.

ediebold commented:
Nope, the change doesn't seem to have helped. Now when I run:

```
ctb-tile -o tiles -f Mesh -C S07E142.hgt
ctb-tile -o tiles -f Mesh -C -l S07E142.hgt
```

I get the following layers.json:

```{
  "tilejson": "2.1.0",
  "name": "S07E142",
  "description": "",
  "version": "1.1.0",
  "format": "quantized-mesh-1.0",
  "attribution": "",
  "schema": "tms",
  "tiles": [ "{z}/{x}/{y}.terrain?v={version}" ],
  "projection": "EPSG:4326",
  "bounds": [ 0.00, -90.00, 180.00, 90.00 ],
  "available": [
    [ { "startX": 1, "startY": 0, "endX": 1, "endY": 0 } ]
   ,[ { "startX": 3, "startY": 0, "endX": 3, "endY": 0 } ]
   ,[ { "startX": 7, "startY": 1, "endX": 7, "endY": 1 } ]
   ,[ { "startX": 14, "startY": 3, "endX": 14, "endY": 3 } ]
   ,[ { "startX": 28, "startY": 7, "endX": 28, "endY": 7 } ]
   ,[ { "startX": 57, "startY": 14, "endX": 57, "endY": 14 } ]
   ,[ { "startX": 114, "startY": 29, "endX": 114, "endY": 29 } ]
   ,[ { "startX": 228, "startY": 59, "endX": 229, "endY": 59 } ]
   ,[  ]
   ,[  ]
   ,[ { "startX": 1832, "startY": 472, "endX": 1833, "endY": 476 } ]
   ,[  ]
  ]
}
```

I also still have the weird repetition on stdout:

`0...10...20...30...40...50...60...70...80...90...0...10...20...30...40...50...60...70...80...90...100 - done.`

ahuarte47 commented:
Ok, thanks I have to study this more quietly. Thanks for your patience.

ahuarte47 commented:
Arggh, it is a silly bug! @ediebold my apologies

ediebold commented:
Thanks @ahuarte47! That seems to have fixed it. I can also confirm that the new cesium friendly option is working great too!

ahuarte47 commented:
I am going to pack some partial commits to leave this pull clean

ahuarte47 committed:
Quantized-mesh output support

Provides two new features:
- New quantized-mesh format ouput
- New metadata file output

ahuarte47 committed:
Fix nodata values

ahuarte47 commented:
Hi @ediebold, I have revert my BuildOverviews experiment, I have not been able to run GDALCreateOverviewDataset properly 😕

petrsloup commented:
I was playing around with this code and it works very nice!

However, I have an issue -- I generated tiles for zoom levels 0-11 (z0 required some manual work as mentioned above), but when I zoom in too much in the Cesium, the terrain completely disappears -- is this known issue? I checked the `layer.json` and the `availability` seems to be correct. Is it somehow possible to force Cesium to "overzoom" the terrain tiles? I tried adding "fake" `availability` levels, but somehow Cesium does not even try to load the more detail tiles (even though the `layer.json` fakes their existence)..

ahuarte47 commented:
Thanks @petrsloup, I have played with my Terrain using only 0-11 levels in Cesium **1.38** (I use the Cesium inspector app) but I have not experimented this issue. Did you ask in Cesium forum?

petrsloup commented:
After further investigation, it turns out the tiles are not always disappearing at the z11. They disappear the z11 in Switzerland, at z9 in Italy and sooner and sooner if I move further west. (No errors in the console.)

I think this might be a dateline issue, because the generated `layer.json` contains `"bounds": [ 0.00, -90.00, 360.00, 90.00 ]`, which is strange.

I reran the process using only part of the data (to avoid the dateline) and everything works correctly (correct bounds and the terrain is not disappearing anywhere).

ahuarte47 committed:
Fix 'Integer Overflow' error when creating tiles of low levels

ahuarte47 commented:
I have changed some of core code to catch the 'Integer overflow' error. CTB builds VRT overviews of input file in order to use when this error appears for low levels.

I would greatly appreciate someone trying their own data

markerikson commented:
@ahuarte47 : my solution for this was just to use `gdalwarp` to create 1/128 and 1/32 versions of the whole-world dataset I'm using, and generate the appropriate zoom levels from those.  I think the 1/128 version gave me zooms 0-7, 1/32 gave me 8-9, and I've been generating 10-14 from the original data.

(Also, this latest set of changes looks like it definitely would cause issues for my fork, as I have it based off an earlier version of your PR. :)  )

ahuarte47 commented:
ctb-tile.cpp, where is the code that saves the tiles, barely changed. I only changed "internal" methods for loading the raster tiles. If this patch definitely fixes this GDAL error, we could generate all levels in an unique step. Could you give it a chance? ;-)

markerikson commented:
I know - just saying that my local fork made changes in `MeshTiler.cpp` and `TerrainTiler.cpp` :)  And since it does what I need it to now, I probably won't be trying to rebase it or merge it compared to your latest changes.

I'll try to throw up a PR today or tomorrow just so the code is visible.

markerikson commented:
@ahuarte47 : out of curiosity, any idea what it would take to implement the "oct-encoded vertex normals" extension for quantized-mesh?

https://cesiumjs.org/data-and-assets/terrain/formats/quantized-mesh-1.0/

ediebold commented:
@markerikson I think if someone got that working, and we had the MBTiles stuff you've been working on, it'd be worth looking at revitalising/redoing the Cesium Terrain Server, so it supports fetching from the SQLite database, and loading the extension data based on the request. This would be something that's much more in my wheelhouse, so I'd be happy to have a crack at it.

ahuarte47 commented:
Hi @markerikson:
>@ahuarte47 : out of curiosity, any idea what it would take to implement the "oct-encoded vertex normals" extension for quantized-mesh?

I did not have any plans to implement this feature, but I found [this](https://github.com/loicgasser/quantized-mesh-tile/blob/master/quantized_mesh_tile/terrain.py#L545) that I think it will be very useful to port to C++

markerikson commented:
I may try experimenting with that library a bit.  Since I've already generated most of my tiles (zooms 0-13 done, halfway through zoom 14 atm), it might be simpler for me to write a Python script that iterates over the tiles in an MBTiles file, decodes them using that library, uses its logic to calculate the vertices and re-encode it, then stuff it back into the MBTiles database.

I'm also trying to push up my CTB branch so I can file a PR and let y'all see the code, but all my connections to Github are timing out today, and I can't tell if it's a Github issue or our corporate proxy being a pain again.  (Probably the latter.)

ahuarte47 commented:
I think I didn't explain myself well, sorry, this line shows the logic in python to add vector-normals, with intention to port it to CTB. It seems easier than to use the library as you say. I will try to spend some time in this.

markerikson commented:
I did understand you :)  The issue is I've already generated 90% of a quantized-mesh tileset, and I don't want to have to regenerate all the tiles from scratch.  

It appears that the `quantized-mesh-tile` Python library has the ability to parse a QM tile, and also potentially calculate those normals from a tile if I put the pieces together right.  So, it would likely be faster for me to try updating the tiles I've already generated to include the normals than it would be to port the logic to C++ and then regenerate the entire tileset.

ahuarte47 commented:
Ah, ok! good luck! :-) but anyway I will study this a bit, it could be interesting

markerikson commented:
I finally had a chance to push up a PR with my MBTiles changes - it's in #61 .  Afraid I don't have time to reconcile these two PRs, but hopefully someone can get that done.

ahuarte47 committed:
Option to write 'Oct-Encoded Per-Vertex Normals' for Terrain Lighting

markerikson commented:
@ahuarte47 : hey, it's great to see you doing that!  I'm actually considering re-generating the set I've been working on, and was figuring I'd have to rebase my changes on top of yours somehow.  Would be nice if we can get all these capabilities functional in one branch within the next day or so :)

ahuarte47 commented:
The most important classes to write a new output format are `CTBOutputStream` and `XXXSerializer`. I have overrided them with `CTBZOutputStream` and `CTBFileTileSerializer` to store "directories of terrain tiles". 

Out of curiosity, why didn't you use the MBTiles support in [GDAL](http://www.gdal.org/frmt_mbtiles.html)?

markerikson commented:
I glanced at the docs for GDAL's MBTiles handling, but it didn't seem like it had the fine-grained support I needed - just the ability to work with a dataset as a whole.

CosmosEcho commented:
Hi @ahuarte47, I got a error in my building. Please help me!
error C2662: 'char **CPLStringList::List(void)' : cannot convert 'this' pointer from 'const CPLStringList' to 'CPLStringList &'
Win10/gdal2.2.4/visual studio 2015

ahuarte47 commented:
Hi @CosmosEcho, I succesfully built with VS2017, but I did a change

CosmosEcho commented:
@ahuarte47 Which gdal version did you used？

ahuarte47 commented:
I use : GDAL 2.3.0dev, released 2017/99/99
Can you build CTB?

CosmosEcho commented:
I can not build. The problem is not solved. I think that I need change my env as yours, but I could not found  GDAL 2.3.0dev https://trac.osgeo.org/gdal/wiki/DownloadSource

ahuarte47 commented:
Is the problem [here](https://github.com/geo-data/cesium-terrain-builder/pull/55/files#diff-849a3b077b2f29b3a30ea2058506c22fR117)?

CosmosEcho commented:
Yes this is.
![image](https://user-images.githubusercontent.com/13760084/38734999-0c410af2-3f5a-11e8-90a4-b426aac40e95.png)

ahuarte47 commented:
It is weird, the signature of this function has not changed for a long time, it is equal in 2.0...[2.2](https://github.com/OSGeo/gdal/blob/release/2.2/gdal/gcore/gdal_priv.h)

Edit:
If you are only going to create Terrain or Mesh Tiles (No GDAL raster outputs, tiff, png,...), you can disable this line to build, this function will not be called.

ahuarte47 committed:
Refactoring of code to easily write new output formats

This defines a set of abstract classes that anyone can override in order
to implement new formats when serializing the terrain tiles.

CosmosEcho commented:
@ahuarte47 OK, thank you！I will try it again.

CosmosEcho commented:
Is that true? "GDALCreateOverviewDataset missing since gdal-2.2"
![image](https://user-images.githubusercontent.com/13760084/38787154-ee3d2228-415e-11e8-8e36-60e9c887d4f0.png)
[https://github.com/OSGeo/gdal/issues/413](url)

CosmosEcho commented:
Hi @ahuarte47, I still can't build completely. Can you give me a builded exe? I want to test my data.

markerikson commented:
I did some experimenting on Friday with re-implementing the MBTiles changes on top of the latest version of this branch, and generated some test terrain that included the vertex normals data.  At first I didn't see any difference visually between quantized-mesh + vertex normals, and quantized-mesh without normals.  I finally figured out that it only matters when you've got `viewer.scene.globe.enableLighting = true`, which changes from a solid light everywhere to simulating globe lighting based on the position of the sun.  

For our use case we don't care about time of day lighting, so I'm going to skip trying to regenerate that terrain set using vertex normals.

I _will_ try to finish re-implementing the MBTiles and missing tiles logic on top of this branch, but it'll have to wait until I have time to finish it outside of work, which may be a while.

ahuarte47 commented:
Hi @CosmosEcho; where could I copy you a zip file (31mb) with CTB+GDAL for Windows x64?

CosmosEcho commented:
@ahuarte47 It's very nice about a zip file. You can send it to my email: 251195335@qq.com. Many thanks!

CosmosEcho commented:
@ahuarte47 if my data problem.
![image](https://user-images.githubusercontent.com/13760084/38908163-921c7720-42f1-11e8-8e4a-3dce78356f95.png)

bjpirt commented:
Hello,
Firstly thanks for adding this feature, it will be great to generate terrain files from GDAL files. Unfortunately I'm having an issue building this branch on Mac OS X and I wondered if you might be able to point me in the right direction to get it compiling. I've installed the GDAL framework (v2.2) that came with QGIS. I'm currently getting this error:

```
/cesium-terrain-builder/src/CTBFileTileSerializer.cpp:117:77: error: member function 'List' not viable: 'this' argument has type
      'const CPLStringList', but function is not marked const
  poDstDS = driver->CreateCopy(temp_filename.c_str(), tile->dataset, FALSE, creationOptions.List(), NULL, NULL);
                                                                            ^~~~~~~~~~~~~~~
/Library/Frameworks/GDAL.framework/Headers/cpl_string.h:522:12: note: 'List' declared here
    char **List() { return papszList; }
           ^
1 error generated.
make[2]: *** [src/CMakeFiles/ctb.dir/CTBFileTileSerializer.cpp.o] Error 1
make[1]: *** [src/CMakeFiles/ctb.dir/all] Error 2
make: *** [all] Error 2
```

Any clues?
Incidentally, the original branch seemed to compile without issues.

ahuarte47 committed:
Try to build in Mac OS X

ahuarte47 commented:
Hello @bjpirt, I have pushed a minor change. Could you test it? thanks!

ahuarte47 commented:
Ups! sorry I did a mesh with git, I accidentally closed this pull, I reopend it as [#64](https://github.com/geo-data/cesium-terrain-builder/pull/64)

Crare commented:
I keep getting this error:
> An error occurred in "T": Failed to obtain terrain tile X: 1 Y: 0 Level: 0 Cesium.js:420

and
> An error occurred in "CesiumTerrainProvider": Failed to obtain terrain tile X: 1 Y: 0 Level: 0. Cesium.js:36877

I've added the files to /0/0/0.terrain and /0/1/0.terrain and I have generated layer.json and tried to add existing layer.json too. I can access the files via http http://localhost/data/tilesets/terrain/test/

And I have same location in CesiumTerrainProvider:
```
let customTerrainProvider = new Cesium.CesiumTerrainProvider({
    url : 'http://localhost/data/tilesets/terrain/test'
});
viewer.scene.terrainProvider = customTerrainProvider;
```

Where is this web.config you want to edit?
where do you put that .htaccess file exactly?
I've tried gzipping the test-folder that has numbered folders and layer-json in it.

I've tried multiple types of tif images and even converting xyz-files but all they get stuck at that same point where I try to show them in Cesium.js.

I'm using docker versions of [ctb](https://github.com/tum-gis/cesium-terrain-builder-docker) and [terrain-server](https://hub.docker.com/r/geodata/cesium-terrain-server/), are they up to date?

BWibo commented:
@Crare I can confirm that the [tumgis/ctb-quantized-mesh](https://github.com/tum-gis/cesium-terrain-builder-docker) Docker image is up-to-date.
It is based on this commit: https://github.com/ahuarte47/cesium-terrain-builder/commit/4bb78ba2e2c13cb0b4631b355cafb1aa94676f65

If you want to make sure you are using the latest commit, you can simply build the image yourself.
The lastest commit from [ahuarte47/cesium-terrain-builder/tree/master-quantized-mesh](https://github.com/ahuarte47/cesium-terrain-builder/tree/master-quantized-mesh) is pulled during the build.
```
git clone https://github.com/tum-gis/cesium-terrain-builder-docker
cd cesium-terrain-builder-docker
docker build -t ctb-quantized-mesh .
```
